src/cli.rs

```diff
@@ -0,0 +1,42 @@
+use clap::*;
+use std::ffi::OsString;
+
+#[derive(Debug, Parser)]
+#[clap(author, version, about, rename_all = "kebab-case")]
+pub struct Cli {
+    /// Run the command on the root workspace project.
+    #[clap(short, long)]
+    pub workspace_root: bool,
+    #[clap(subcommand)]
+    pub command: Option<Command>,
+}
+
+#[derive(Debug, Subcommand)]
+#[clap(rename_all = "kebab-case")]
+pub enum Command {
+    #[clap(alias = "run-script")]
+    Run(RunArgs),
+    #[clap(alias = "i")]
+    Install(PassedThroughArgs),
+    #[clap(alias = "up")]
+    Update(PassedThroughArgs),
+    #[clap(external_subcommand)]
+    Other(Vec<String>),
+}
```

Comment:
I copy-pasted some help messages from `pnpm`.

```suggestion
#[derive(Debug, Subcommand)]
#[clap(rename_all = "kebab-case")]
pub enum Command {
    /// Runs a defined package script. 
    #[clap(alias = "run-script")]
    Run(RunArgs),
    /// Installs all dependencies of the project in the current working directory.
    /// When executed inside a workspace, installs all dependencies of all projects.
    #[clap(alias = "i")]
    Install(PassedThroughArgs),
    /// Updates packages to their latest version based on the specified range.
    /// You can use "*" in package name to update all packages with the same pattern.
    #[clap(alias = "up")]
    Update(PassedThroughArgs),
    /// Other commands to pass to pnpm.
    #[clap(external_subcommand)]
    Other(Vec<String>),
}
```

Reply:
> /// Other commands to pass to pnpm.

No, actually now only i and up are passed to pnpm
everything else is the same as running `pnpm exec args...`

So I am not sure what should be done. We should probably list all the pnpm commands. However, the matcher currently looks ugly. Maybe there's a way to improve it.

Reply:
You intended `pn` as a wrapper for `pnpm`, right? Or you want to eventually replace JavaScript `pnpm` with this? Either way, I think it makes sense to fallback to `pnpm` for command `pn` hasn't implement yet.

> the matcher currently looks ugly

What is this "matcher" you are referring to?

Reply:
The plan is currently to use this wrapper for running scripts. As it is much faster. It also helps me to learn rust.

Then I will try to use node.js bindings with node-rs. I want to do fetching of packages, extraction and writing to the content-addressable store using rust. I hope this single change can make pnpm faster a few times.

Reply:
> I want to do fetching of packages, extraction and writing to the content-addressable store using rust.

If Rust does both running scripts and installing packages (according to your plan), then I think it pretty much replaces JavaScript `pnpm` already. Is there still a need to create a Node.js binding? Unless there are some features in `pnpm` that is too big to rewrite in Rust.

Reply:
This is just a small part of pnpm's codebase, so it won't replace pnpm. A full rewrite would be very hard and time consuming. Maybe it will happen in the future but we don't have the resources to do it now. And we cannot just switch all our efforts to it.

The goal is to make pnpm fast enough for users to don't care about the alternatives that are fully written in rust or zig.

Reply:
`pn` as it is currently is very small indeed. But you also said "I want to do fetching of packages, extraction and writing to the content-addressable store using rust" which is pretty much the biggest feature of `pnpm`.

Reply:
no, it is not the biggest feature. It is a relatively small independent part of pnpm. Actually, stackblitz has already done this. They run pnpm with their own custom fetcher and it is much faster.

The tarball fetcher part is basically just two or three packages from the pnpm monorepo.

Reply:
also, I am not opposed to a full rewrite. But I want improvements as soon and I want them to be useful and production ready for most of our users. So rewriting a relatively small, isolated part that would give a significant boost to the performance is the best solution I believe.